File: modules/zooloo-printer/android/src/main/java/expo/modules/zoolooprinter/ZoolooPrinterModule.kt
package expo.modules.zoolooprinter

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ZoolooPrinterModule : Module() {
  private val printerManager = PrinterManager()

  override fun definition() = ModuleDefinition {
    Name("ZoolooPrinter")

    AsyncFunction("getPairedDevices") { ->
      printerManager.getPairedDevices()
    }

    AsyncFunction("connect") { macAddress: String ->
      printerManager.connect(macAddress)
    }

    AsyncFunction("disconnect") { ->
      printerManager.disconnect()
    }

    AsyncFunction("printText") { text: String ->
      printerManager.printText(text)
    }
    
    // Commands in ESC/POS can be passed as an array of numbers
    AsyncFunction("printCommand") { command: List<Int> ->
      val bytes = command.map { it.toByte() }.toByteArray()
      printerManager.printCommand(bytes)
    }
  }
}

File: modules/zooloo-printer/src/ZoolooPrinterModule.ts
import { NativeModule, requireNativeModule } from 'expo';

export type BluetoothDevice = {
  name: string;
  macAddress: string;
};

declare class ZoolooPrinterModule extends NativeModule {
  getPairedDevices(): Promise<BluetoothDevice[]>;
  connect(macAddress: string): Promise<boolean>;
  disconnect(): Promise<boolean>;
  printText(text: string): Promise<boolean>;
  printCommand(command: number[]): Promise<boolean>;
}

// This call loads the native module object from the JSI.
export default requireNativeModule<ZoolooPrinterModule>('ZoolooPrinter');
